File: app/src/main/java/dev/ragnarok/fenrir/api/model/VkApiAttachments.java
package dev.ragnarok.fenrir.api.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class VkApiAttachments {

    public ArrayList<Entry> entries;

    public VkApiAttachments() {
        entries = new ArrayList<>(1);
    }

    public VkApiAttachments(int initialSize) {
        entries = new ArrayList<>(initialSize);
    }

    public List<Entry> entryList() {
        return Collections.unmodifiableList(entries);
    }

    public void append(VKApiAttachment attachment) {
        if (entries == null) {
            entries = new ArrayList<>(1);
        }

        entries.add(new Entry(attachment.getType(), attachment));
    }

    public void clear() {
        if (entries != null) {
            entries.clear();
        }
    }

    public void append(List<? extends VKApiAttachment> data) {
        for (VKApiAttachment attachment : data) {
            append(attachment);
        }
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    public boolean nonEmpty() {
        return size() > 0;
    }

    public int size() {
        return entries == null ? 0 : entries.size();
    }

    public static class Entry {

        public final String type;
        public final VKApiAttachment attachment;

        public Entry(String type, VKApiAttachment attachment) {
            this.type = type;
            this.attachment = attachment;
        }
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/api/model/VKApiAttachments.kt
package dev.ragnarok.fenrir.api.model

import dev.ragnarok.fenrir.api.adapters.AttachmentsDtoAdapter
import dev.ragnarok.fenrir.api.adapters.AttachmentsEntryDtoAdapter
import kotlinx.serialization.Serializable
import java.util.*

@Serializable(with = AttachmentsDtoAdapter::class)
class VKApiAttachments {
    val entries: ArrayList<Entry>

    constructor() {
        entries = ArrayList(1)
    }

    constructor(initialSize: Int) {
        entries = ArrayList(initialSize)
    }

    fun entryList(): List<Entry> {
        return Collections.unmodifiableList(entries)
    }

    fun append(attachment: VKApiAttachment) {
        entries.add(Entry(attachment.getType(), attachment))
    }

    fun clear() {
        entries.clear()
    }

    fun append(data: List<VKApiAttachment>) {
        for (attachment in data) {
            append(attachment)
        }
    }

    val isEmpty: Boolean
        get() = size() <= 0

    fun nonEmpty(): Boolean {
        return size() > 0
    }

    fun size(): Int {
        return entries.size
    }

    @Serializable(with = AttachmentsEntryDtoAdapter::class)
    class Entry(val type: String, val attachment: VKApiAttachment)
}